### PR TITLE
Cleanup stale buckets

### LINKED ISSUE #1295: Investigate large number of buckets in bucket directory
Currently all bucket files are stored in 1 flat directory:
```
/buckets
    bucket-002d4e82245617c72c5eedbdd7ebda6a76d0ccf4bc714fd3eccb57e4cd7dd5e1.xdr  
    bucket-54789b62f64896e05aa457d743e9a4a08ed3f665a21a56c578955fd646e521c4.xdr  
    bucket-a986b1bcc9fd737ab2dd529b86db957ada76fcb66bac758dbfb19c76fe242bbf.xdr
    bucket-004106780e8ff5466dbaa6b99a95f373587abc176808c0c9b26194ed47da54cf.xdr
```
On most filesystems (except maybe ext4) it causes performance issues when you have 10,000 files and more. It just blows away my terminal when I'm trying to list files or calculate storage.

We could use subdirectories to improve situation, similar to what varnish is doing:
```
/buckets/0/0/2/d4e82245617c72c5eedbdd7ebda6a76d0ccf4bc714fd3eccb57e4cd7dd5e1.xdr  
/buckets/0/0/4/106780e8ff5466dbaa6b99a95f373587abc176808c0c9b26194ed47da54cf.xdr
/buckets/5/4/7/89b62f64896e05aa457d743e9a4a08ed3f665a21a56c578955fd646e521c4.xdr  
/buckets/a/9/8/6b1bcc9fd737ab2dd529b86db957ada76fcb66bac758dbfb19c76fe242bbf.xdr
```

### PR DESCRIPTION
**Pull request by vogel**
Fixes #1295

### CODE REVIEW

File: src/bucket/BucketManagerImpl.cpp [hunk 1]
```diff
@@ -240,6 +240,48 @@ BucketManagerImpl::forgetUnreferencedBuckets()
...
+    return referenced;
+}
+
+void
+BucketManagerImpl::cleanupStaleFiles()
+{
+    std::lock_guard<std::recursive_mutex> lock(mBucketMutex);
+    auto referenced = getReferencedBuckets();
+    std::transform(std::begin(mSharedBuckets), std::end(mSharedBuckets),
+                   std::inserter(referenced, std::end(referenced)),
```
Comment (MonsieurNicolas) [inline src/bucket/BucketManagerImpl.cpp]: `std::back_inserter` is what you want here

File: src/util/Fs.h [hunk 2]
```diff
@@ -40,6 +42,11 @@ bool mkdir(std::string const& path);
 // Make a dir path like mkdir -p, i.e. recursive, uses '/' as dir separator
 bool mkpath(std::string const& path);
 
+// Get list of all files with names matching predicate
```
Comment (MonsieurNicolas) [inline src/util/Fs.h]: add comment that file names are relative to path

File: src/bucket/BucketManagerImpl.cpp (hunk 1 above)
Comment (vogel) [reply src/bucket/BucketManagerImpl.cpp]: It does not work on std::set, so I'm using regular inserter

**Comment by vogel** [comment]
Updated.

**Comment by MonsieurNicolas** [comment]
I have a commit ready to add to this pr (Windows support) but I don't have rights to push to your branch @vogel

**Comment by MonsieurNicolas** [comment]
I'm wondering if we should delete buckets that are older than let's say a day to avoid re-downloading buckets for no reason between aborted catchup. Or are those buckets in a temp folder anyways?

**Comment by MonsieurNicolas** [comment]
@vogel see my question above

**Comment by vogel** [comment]
I don't think I understand that question.

We are deleting buckets that we are not referencing to, so why should we delete all buckets older than one day? Some buckets change more rarely than once a day and we should keep them.

Which buckets do you have in mind?

**Comment by MonsieurNicolas** [comment]
yes, what I am asking is if we should keep buckets that are more recent than a day in the event that we restart stellar core during catchup to avoid re-downloading a bunch of stuff from history. If we place those buckets in temp folders, then this won't change anything, but I am not sure without going through the code.

**Comment by vogel** [comment]
It uses tmp dir, see `BucketDownloadWork::BucketDownloadWork`.

**Comment by MonsieurNicolas** [comment]
yeah that's what I suspected. Thanks for looking

**Comment by MonsieurNicolas** [comment]
r+ 2aecca8f95bf9c8441bc9b0e1851231d0fa81a3b

**Comment by MonsieurNicolas** [comment]
@latobarita: retry

**Comment by MonsieurNicolas** [comment]
@latobarita: retry

**Pull request merged**